tests/helpers/linux/test_pid.py

```diff
@@ -32,3 +32,13 @@ def test_find_task(self):
     def test_for_each_task(self):
         pid = os.getpid()
         self.assertTrue(any(task.pid == pid for task in for_each_task(self.prog)))
+
+    def test_for_each_task_prog_threads(self):
```

Comment:
@osandov
There's a bit of a problem here I could use your input on.

So this test is presently implemented by `zip`'ing the two iterators because `zip` will stop producing values once the shortest of the two iterators has been exhausted. This is desirable because for the live kernel, as the current set of tasks may change during the processing of these iterators such that one has more tasks than the other. From the many runs I have performed while testing, this _almost_ always works, but _very rarely_ it fails. I'm assuming the failures occur when we get unlucky enough for the underlying set of tasks to change at such a time that both `self.prog.threads()` and `for_each_task(self.prog)` have the same number of tasks, but the last task (or one close to being the last) is different. 

Given your knowledge of the Kernel, is there any way to "freeze" the relevant state here so that we could avoid this issue? One simplistic idea I had was just running this test on a kernel core dump and not on the live kernel (since the implementations share almost all of their code), but the issue is that kernel core dumps (at least the ones I've seen) are _huge_, even when compressed, making it unrealistic to store them in the repo as I did for the userspace core dump test. I suppose we could upload a kernel core dump as a GitHub releases asset much like you did for the various kernel versions, or (and not sure if this would work) perhaps we could get a smaller core dump by causing a crash on an Alpine Linux VM or similarly light-weight distro and using that (I'm assuming that having a much smaller set of running processes would result in a correspondingly smaller kernel core dump).

Reply:
Resolved on video call. The conclusion is as follows:

First of all, testing `for_each_task` against `prog.threads()` isn't terribly useful, since they rely on the same internal implementation. But to address the larger question of how to test the iterator functionality in general, this will depend on which target we are testing:

**Live kernel**
We can test `for_each_task` by first spawning a bunch of threads (this can be done directly from Python) and noting down their `tid`s, then checking that all of them are represented in the iterator returned by `for_each_task`.

**Kernel core dump**
Within the QEMU VM, we first spawn a bunch of threads (the same way as for the live kernel test), and save their `tid`s to a file on disk. We then deliberately cause a kernel panic (i.e. `echo c > /proc/sysrq-trigger`), and assuming `kdump` was set up correctly once we boot into the capture kernel, we can run `drgn` on the crash dump (i.e. `/proc/vmcore`), and assert that the iterator returned by `for_each_task` contains all of the `tid`s that we saved.

**Userspace core dump**
This has already been implemented, see [`tests/test_program.py:test_threads`](https://github.com/osandov/drgn/pull/129/files#diff-be39934161cfc173a34e22536cf23d3f8668a11da20e4f1951b4f47497884587R833), and is accomplished by testing against a static userspace core dump with known contents.